### ISSUE TITLE
Unable to Preview files generated from partials using preprocessors

### ISSUE THREAD

**Issue by thekip** [timo-reymann/intellij-mjml-support#111]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Used versions

* Host OS: Linux/Darwin/Windows
* IDE: <Name> <Version>

## Problem description
Imagine the following structure:

```
 - _wrapper.mjml
 - signin.mjml
 - register.mjml
 - etc
```
Where `signin` and `register` templates use `_wrapper.mjml` as base layout. 
We are using [Eta layouts](https://eta.js.org/docs/learn/layouts) for this purpose. 

So the `wrapper.mjml` looks like this: 

```html
<mjml>
  <mj-head>
  </mj-head>
 <mj-body>
  <!-- some common layout -->
 <%~ it.body %>
  <!-- some common layout -->
 </mj-body>
</mjml>
```

`signin.mjml` and `register.mjml` looks like

```html
<% layout('./wrapper.mjml') %>

<mj-section>
  <mj-column>content</mj-column>
</mj-section>
```

I implemented custom `intellij_render.js` file where integrated Eta engine into workflow. But problem is, intelij plugin give no chance to this custom renderer to be executed. Because it checks the validity of file before render function is called. So i always see an error message in preview `Ensure file has a mjml root tag and at least a mj-body child`.

Corresponding code is stored here /mjml_support/editor/ui/MjmlPreviewFileEditor.kt:184 in `updateHtml` function. It first checks if mjml file is valid by simply checking are root and body presented. But they are not, because they come from wrapper using Eta. 

## Expected Behavior

I'm expecting the plugin would either always use renderer function, or make another extension point which would return mjml (not html) to be validated. Something like:

`intellij_render.js`

```js
module.exports = {
  // you can override only specific functions. the rest would be taken from bundled script
  // what would be returned from this function would be used for validating and etc.   
  before: function(...args) {}
  renderToHtml: function(...args) {}
  after: function(...args) {}
}
```

The current behaviour also closes the door for such awesome template engines as Pug which has completely different syntax.

**Comment by thekip** [comment]
## Update

Checked MJML desktop application, they allow to edit and preview partials (files without root and body tags). They achieve this by implicitly adding those tags and passing result to the mjml2html function. 

Because of that and because of support for `.mjmlconfig.js/preprocessors` it's possible to implement any kind of  preprocess flow in app. 

But i would like to have the same ability in IDE as well. So the best option i see is to align this plugin with app behaviour.

**Comment by timo-reymann** [comment]
Heya @thekip,

thank you for the very detailed issue, that makes my life a lot easier! :)

To your solution approaches:

**another extension point:**
This is not so easy, since the node.js script is only called as a black box from IntelliJ so we dont have something like partial overrides or Node.js functionality, and i want to keep this part as simple as possible.

**implictly adding the tags**
Especially for your use case this would result in an rendering error from my understanding. The MJML file contains the eta layout part and the processing script then would have nested mjml file structures. One from the implict added and one from the templating with eta.

Would it be sufficient for your use case to simply allow disabling the validation?
It is shareable via VCS so you could check in the MJML settings including the "validation disable option"

Kind regards
Timo

**Comment by thekip** [comment]
> Especially for your use case this would result in an rendering error from my understanding. The MJML file contains the eta layout part and the processing script then would have nested mjml file structures. One from the implict added and one from the templating with eta.

Agree, that's exactly what i  faced with native app and fixed with quite ugly code like this: 

```js
  preprocessors: [
    (mjml) => {
      if (mjml.includes('<% layout(')) {

        mjml = mjml
          .replace('<mjml>', '')
          .replace('</mjml>', '')
          .replace('<mj-body>', '')
          .replace('</mj-body>', '');
      }
```

So you proposal to just disable validation from settings would be nice and fix the issue.

**Comment by timo-reymann** [comment]
Heya,
will be available with the next release :)

Preview:
![image](https://user-images.githubusercontent.com/21024948/158249494-78d7a290-7e52-43d2-a2df-52a110c75075.png)

Settings:
![image](https://user-images.githubusercontent.com/21024948/158249717-6eeb2683-d3e0-4d08-a3b5-63b0f607f866.png)

**Comment by timo-reymann** [comment]
[NEW SNAPSHOT]
Latest snapshot mentioning this issue has been published to marketplace:
    <a href="https://plugins.jetbrains.com/plugin/16418-mjml-support/versions/snapshot">2022.03.14-600</a>

**Comment by thekip** [comment]
Great Job! Installed latests snapshot and was able to setup preview for our pre rendering workflow. Huge step in development experience. Thanks.

**Comment by timo-reymann** [comment]
Will be included in 2.1.0
